File: src/Agent/Agent.Core/Models/Api/v1/TodoPlan.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

using System.Text.Json.Serialization;

namespace Agent.Core.Models.Api.v1;

public sealed record TodoPlan
{
    public required Guid Id { get; set; }
    public required string Title { get; set; }
    public required Guid ThreadId { get; set; }
    public required Guid TriggerMessageId { get; set; }

    [JsonConverter(typeof(JsonStringEnumConverter))]
    public required TodoPlanStatus Status { get; set; }

    public required IReadOnlyList<TodoItem> Items { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime? LastUpdated { get; set; }
}

File: src/Agent/Agent.Web/Client/src/src/Common/Contracts/DataPlane/TodoPlan.ts
export interface TodoPlan {
    id: string;
    title: string;
    threadId: string;
    triggerMessageId: string;
    status: TodoPlanStatus;
    items: TodoItem[];
    createdAt: string;
    lastUpdated?: string;
}

export interface TodoItem {
    content: string;
    activeForm: string;
    status: TodoItemStatus;
    order: number;
    startedAt?: string;
    completedAt?: string;
}

export enum TodoPlanStatus {
    Planning = 'Planning',
    InProgress = 'InProgress',
    Completed = 'Completed',
    Cancelled = 'Cancelled',
}

export enum TodoItemStatus {
    Pending = 'Pending',
    InProgress = 'InProgress',
    Completed = 'Completed',
    Failed = 'Failed',
}
